### PR TITLE
Add GreeterRow + Accordion.Link component

### PR DESCRIPTION
**Pull request by ItsJonQ**
### GreeterRow

screen recording 2019-03-06 at 11 43 am

This update adds a new `GreeterRow` component. It's built with the new `Accordion.Link` component.

### Accordion.Link

screen recording 2019-03-05 at 02 18 pm

This update refactors and enhances the `Accordion` sub component to support Link rendering, with either `to` or `href`. This allows for an Accordion items UI and logic to be recycled to support React Router navigation.

### Refactors

All of `Accordion` was refactored.

1. To convert things to the TypeScript and latest/greatest conventions
2. Replacing the prop-waterfall method with `PropProvider`
3. Updating styles to match latest HSDS: Product UI

This allowed for a cleaner implementation of link logic to be inserted into `Accordion.Section` and `Accordion.Title`

### propConnect: mapPropConnectedProps

This update also enhances the `propConnect()` HOC with a new option, `mapPropConnectedProps`. It works identically to the `mapStateToProps` pattern from React Redux'` `connect()` HOC. This allows for finer grain control when constructing and composing components.

---

## QA/Testing

cc'ing @ryan-mulrooney for the initial `Accordion.Link` QA

Demo Link:

https://deploy-preview-512--hsds-react.netlify.com/?selectedKind=GreeterRow&selectedStory=Default&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

### CODE REVIEW

**Comment by ItsJonQ** [comment]
## Deploy Preview Link:

https://deploy-preview-512--hsds-react.netlify.com/?selectedKind=Accordion&selectedStory=Row%20%28Link%29&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

**Comment by ItsJonQ** [comment]
Taking off the WIP label!

@digitaldesigner + @ryan-mulrooney , ready for review!

https://deploy-preview-512--hsds-react.netlify.com/?selectedKind=GreeterRow&selectedStory=Default&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

**Comment by ItsJonQ** [comment]
Tagging @knicklabs for code review!

Also @jmauerhan , in case you're interested ;)

File: src/components/Icon/styles/Icon.css.js [hunk 1]
```diff
@@ -125,4 +127,19 @@ function makeStateColorStyles(): string {
   )
 }
 
+function adjust24To20Scale(): string {
+  return `
+    &[height="24"],
+    &[width="24"],
+    &[viewBox*="24"] {
+      height: calc(24/20 * 100%);
```
Comment (ItsJonQ) [inline src/components/Icon/styles/Icon.css.js]: Magic SVG -> CSS MATHS

File: src/components/Accordion/Accordion.tsx [hunk 2]
```diff
@@ -141,29 +140,35 @@ class Accordion extends PureComponent<AccordionProps, AccordionState> {
...
+    const componentClassName = getComponentClassName(this.props)
 
     return (
       <AccordionUI
         {...getValidProps(rest)}
         className={componentClassName}
         role="tablist"
       >
-        {content}
+        <PropProvider value={this.getPropProviderProps()}>
```
Comment (ItsJonQ) [inline src/components/Accordion/Accordion.tsx]: No more `React.Children` cloning + prop waterfalling. We usin' `PropProvider`!

File: src/components/Accordion/Accordion.Link.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+// Higher-order wrapper to render an enhanced Accordion.Section and
+// Accordion.Title
+export const Link = props => {
+  const { children, isSeamless: isSeamlessProp, to, href, ...rest } = props
+  const isLink = !!(to || href)
+  const isSeamless = isLink ? false : isSeamlessProp
+  const isOpen = false
+
+  const sectionProps = {
+    ...rest,
```
Comment (jmauerhan) [inline src/components/Accordion/Accordion.Link.tsx]: What is `...rest` ?

Comment (ItsJonQ) [reply src/components/Accordion/Accordion.Link.tsx]: @jmauerhan It's a way to [destructure](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) an Array or Object in JS.

```js
// Original object
const obj = {
  a: 1,
  b: 2,
  c: 3
}

// Plucking out something, and assigning new const (variables)
const { a, ...rest } = obj

// const a = 1
// const rest = {
//   b: 2,
//   c: 3
// }
```

**Comment by ItsJonQ** [comment]
Updated Icon CSS to use `transform` for resize MATH. Thanks for the idea @plbabin

File: src/components/Accordion/Accordion.Link.tsx (hunk 3 above)
Comment (ryan-mulrooney) [reply src/components/Accordion/Accordion.Link.tsx]: Thanks for that explanation :)

**Comment by ItsJonQ** [comment]
@ryan-mulrooney Both amazing catches! Thank you!!

**Comment by ryan-mulrooney** [comment]
> @ryan-mulrooney Both amazing catches! Thank you!!

Haha, GitHub stripped ma sh1t!

**Comment by ItsJonQ** [comment]
@ryan-mulrooney Fixed!

screen shot 2019-03-08 at 10 11 26 am

screen shot 2019-03-08 at 10 31 50 am

**Comment by ItsJonQ** [comment]
screen shot 2019-03-08 at 11 17 56 am

File: src/utilities/component.ts [hunk 4]
```diff
@@ -183,3 +183,20 @@ export const renderRenderPropComponent = (
   }
   return null
 }
+
+export const renderChildrenSafely = (
```
Comment (ItsJonQ) [inline src/utilities/component.ts]: Will abstract this out to `@helpscout/react-utils` in the future

**Comment by ryan-mulrooney** [comment]
LGTM now! 👏

**Comment by ItsJonQ** [comment]
Thanks for the review all! Gonna merge/release now

**Pull request merged**